### PR TITLE
Transaction guardrails

### LINKED ISSUE #11483: sql: add txn rows written/read guardrails
PR: https://github.com/cockroachdb/cockroach/pull/69202

From release notes:
> The new [cluster settings](../v21.2/cluster-settings.html) `sql.defaults.transaction_rows_written_log`, `sql.defaults.transaction_rows_written_err`, `sql.defaults.transaction_rows_read_log`, and `sql.defaults.transaction_rows_read_err` (as well as the corresponding session variables) have been introduced. These settings determine the "size" of the transactions in written and read rows upon reaching of which the transactions are logged or rejected. The logging will go into the [`SQL_PERF`](../v21.2/logging.html#sql_perf) logging channel. Note that the internal queries (i.e., those issued by CockroachDB internally) cannot error out but can be logged instead into [`SQL_INTERNAL_PERF`](../v21.2/logging.html#sql_internal_perf) logging channel. The "written" limits apply to [`INSERT`](../v21.2/insert.html), [`INSERT INTO SELECT FROM`](../v21.2/insert.html), [`INSERT ON CONFLICT`](../v21.2/insert.html), [`UPSERT`](../v21.2/upsert.html), [`UPDATE`](../v21.2/update.html), and [`DELETE`](../v21.2/delete.html) whereas the "read" limits apply to [`SELECT`](../v21.2/select-clause.html) statement in addition to all of these. These limits will not apply to [`CREATE TABLE AS`](../v21.2/create-table.html), [`SELECT`](../v21.2/select-clause.html), [`IMPORT`](../v21.2/import.html), [`TRUNCATE`](../v21.2/truncate.html), [`DROP`](../v21.2/drop-table.html), [`ALTER TABLE`](../v21.2/alter-table.html), [`BACKUP`](../v21.2/backup.html), [`RESTORE`](../v21.2/restore.html), or [`CREATE STATISTICS`](../v21.2/create-statistics.html) statements. Note that enabling `transaction_rows_read_err` guardrail comes at the cost of disabling the usage of the auto commit optimization for the mutation statements in implicit transactions. [#69202][#69202]

### LINKED ISSUE #11651: sql: adjust recently added logging events for txn row count guardrails
https://github.com/cockroachdb/cockroach/pull/70014 --- Release note (ops change): the meaning of the recently introduced `transaction_rows_written_err` and `transaction_rows_read_err` (as well as the corresponding `_log` variables) have been adjusted a bit to indicate the largest number of rows that is still allowed. In other words, originally reaching the limit would result in an error, and now only exceeding the limit would.  Release justification: low-risk adjustment to new functionality.

### PR DESCRIPTION
**Pull request by stbof**
Fixes #11483.
Fixes #11651.

Related to #12247.

### CODE REVIEW

**Review by yuzefovich — APPROVED** [review]
:lgtm:

Note that the guardrails (these "rows written" and "rows read" as well as "max_row_size" have been backported to 21.1 release and are available from 21.1.11). It might make sense to make the adjustment to 21.1 docs right away too.

> Reviewed 1 of 1 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/12212#-:-Mo5UiGe-hY4WLyCkYoQ:b-hc9yca)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ianjevans, @stbof, and @vy-ton)

___
*[v21.2/transactions.md, line 268 at r1](https://reviewable.io/reviews/cockroachdb/docs/12212#-Mo5UVDH-sIfLc-1r9WT:-Mo5UVDH-sIfLc-1r9WU:bg6lu8y) ([raw file](https://github.com/cockroachdb/docs/blob/ce34199276c50afc1c5096375b8efc4613452cd4/v21.2/transactions.md#L268)):*
> ```Markdown
> ## Limit the number of rows written or read in a transaction
> 
> You can configure cluster and sessions to limit the number of rows written or read in a transaction. This allows you configure CockroachDB to log or reject statements that could destabilize a cluster or violate application best practices.
> ```

nit: I'm not sure whether it is intentional and whether it is important (I don't this on `.md` page), but some lines/words here seem to be italicized.

File: v21.2/transactions.md [hunk 1]
```diff
@@ -263,6 +263,27 @@ The CockroachDB `SERIALIZABLE` level is stronger than the ANSI SQL `READ UNCOMMI
 
 For more information about the relationship between these levels, see [this paper](http://arxiv.org/ftp/cs/papers/0701/0701157.pdf).
 
+## Limit the number of rows written or read in a transaction
+
+You can configure cluster and sessions to limit the number of rows written or read in a transaction. This allows you configure CockroachDB to log or reject statements that could destabilize a cluster or violate application best practices.
```
Comment (ianjevans) [inline v21.2/transactions.md:268]:
```suggestion
You can limit the number of rows written or read in a transaction at the cluster or session level. This allows you configure CockroachDB to log or reject statements that could destabilize a cluster or violate application best practices.
```

**Review by ianjevans — COMMENTED** [review]
LGTM, except for that one sentence.

**Review by vy-ton — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/12212)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @stbof and @yuzefovich)

___
*[v21.2/transactions.md, line 270 at r2](https://reviewable.io/reviews/cockroachdb/docs/12212#-MoEw_zh2qmxfhZHne9F:-MoEw_zh2qmxfhZHne9G:bdp4fd2) ([raw file](https://github.com/cockroachdb/docs/blob/21aa808d413226b3d23686a9307643de57f5a572/v21.2/transactions.md#L270)):*
> ```Markdown
> You can limit the number of rows written or read in a transaction at the cluster or session level. This allows you configure CockroachDB to log or reject statements that could destabilize a cluster or violate application best practices.
> 
> Use the [cluster](cluster-settings.html) and [session](set-vars.html) settings `sql.defaults.transaction_rows_written_log`,
> ```

I would reference the session setting names here, since that will likely be the preferred pattern for most developers. 
___
*[v21.2/transactions.md, line 284 at r2](https://reviewable.io/reviews/cockroachdb/docs/12212#-MoEw0TO6_vMqcp9Y1_Z:-MoEw0TP6WkRT9XCTmfG:bcd4loh) ([raw file](https://github.com/cockroachdb/docs/blob/21aa808d413226b3d23686a9307643de57f5a572/v21.2/transactions.md#L284)):*
> ```Markdown
> 
> {{site.data.alerts.callout_info}}
> Enabling `transaction_rows_read_err` disables the auto commit optimization for mutation statements in implicit transactions. For write limits CockroachDB can count how many rows have been modified before using the auto commit optimization. However, for read limits CockroachDB doesn't have that information on the write path, so must disable the auto commit.
> ```

@yuzefovich Is this a performance optimization? What's the "so what" for a user?

**Review by yuzefovich — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/12212)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @stbof, @vy-ton, and @yuzefovich)

___
*[v21.2/transactions.md, line 284 at r2](https://reviewable.io/reviews/cockroachdb/docs/12212#-MoEw0TO6_vMqcp9Y1_Z:-MoF1wSi-SH8tcuXPtJe:b-4vpnwb) ([raw file](https://github.com/cockroachdb/docs/blob/21aa808d413226b3d23686a9307643de57f5a572/v21.2/transactions.md#L284)):*
<details><summary><i>Previously, vy-ton (Vy Ton) wrote…</i></summary><blockquote>

@yuzefovich Is this a performance optimization? What's the "so what" for a user?
</blockquote></details>

Yes, it's a performance optimization. I did a quick search in our docs, and I don't think we mention it there.

Here’s what it means:
- CRDB can execute transactions in two “ways”
  - an explicit transaction: this will be BEGIN statement followed by some number of other statements (like SELECT, INSERT, etc) followed by COMMIT statement. In this scenario the transaction is committed on COMMIT. At the minimum this kind of setup requires 3 round-trips from the client to CRDB server: 1 for BEGIN, 1 for some meaningful statement, 1 for COMMIT
  - an implicit transaction: if you don’t use an explicit BEGIN - COMMIT pair, then each statement runs as a separate transaction. So if you simply do SELECT …, then it’ll have only one round-trip between client and server.

Now, for implicit transactions, if we have a mutation statement (like INSERT, UPDATE, etc), we have an “auto commit optimization” which allows us to perform the KV level commit together with the last KV batch that contains the mutations.

Imagine that we want to write 3 rows, with something like `INSERT INTO t VALUES (1), (2), (3)`. Without auto commit we would have two KV-level operations: one for writing 3 KV pairs, and another for KV-level commit. Each KV-level operation might require a network hop to a remote node. However, with auto commit optimization we do a single KV-level operation (which includes writing 3 KV pairs as well as the KV-level commit).

**Comment by stbof** [comment]
___
*[v21.2/transactions.md, line 270 at r2](https://reviewable.io/reviews/cockroachdb/docs/12212#-MoEw_zh2qmxfhZHne9F:-MoFEfRTDj6dAg1F1QKW:b-5t9l7l) ([raw file](https://github.com/cockroachdb/docs/blob/21aa808d413226b3d23686a9307643de57f5a572/v21.2/transactions.md#L270)):*
<details><summary><i>Previously, vy-ton (Vy Ton) wrote…</i></summary><blockquote>

I would reference the session setting names here, since that will likely be the preferred pattern for most developers. 
</blockquote></details>

I'm not sure I understand what you are asking for. The setting names are the same for clusters and sessions. Should I drop cluster settings?

**Comment by stbof** [comment]
___
*[v21.2/transactions.md, line 284 at r2](https://reviewable.io/reviews/cockroachdb/docs/12212#-MoEw0TO6_vMqcp9Y1_Z:-MoFF9GYBWLJnmfFJMJ6:b-6caoqm) ([raw file](https://github.com/cockroachdb/docs/blob/21aa808d413226b3d23686a9307643de57f5a572/v21.2/transactions.md#L284)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Yes, it's a performance optimization. I did a quick search in our docs, and I don't think we mention it there.

Here’s what it means:
- CRDB can execute transactions in two “ways”
  - an explicit transaction: this will be BEGIN statement followed by some number of other statements (like SELECT, INSERT, etc) followed by COMMIT statement. In this scenario the transaction is committed on COMMIT. At the minimum this kind of setup requires 3 round-trips from the client to CRDB server: 1 for BEGIN, 1 for some meaningful statement, 1 for COMMIT
  - an implicit transaction: if you don’t use an explicit BEGIN - COMMIT pair, then each statement runs as a separate transaction. So if you simply do SELECT …, then it’ll have only one round-trip between client and server.

Now, for implicit transactions, if we have a mutation statement (like INSERT, UPDATE, etc), we have an “auto commit optimization” which allows us to perform the KV level commit together with the last KV batch that contains the mutations.

Imagine that we want to write 3 rows, with something like `INSERT INTO t VALUES (1), (2), (3)`. Without auto commit we would have two KV-level operations: one for writing 3 KV pairs, and another for KV-level commit. Each KV-level operation might require a network hop to a remote node. However, with auto commit optimization we do a single KV-level operation (which includes writing 3 KV pairs as well as the KV-level commit).
</blockquote></details>

So should my actions be: 

Should I add a new section to the transactions page that describes explicit and implicit transactions and describes the auto commit behavior for the latter?
Amend the guardrails section to say: Enabling `transaction_rows_read_err` disables the auto commit _performance_ optimization and link to the new section?

**Review by stbof — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/12212)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @stbof, @vy-ton, and @yuzefovich)

___
*[v21.2/transactions.md, line 268 at r1](https://reviewable.io/reviews/cockroachdb/docs/12212#-Mo5UVDH-sIfLc-1r9WT:-Mo63cSE0uu6Zy9k5UO2:bbbjx4b) ([raw file](https://github.com/cockroachdb/docs/blob/ce34199276c50afc1c5096375b8efc4613452cd4/v21.2/transactions.md#L268)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: I'm not sure whether it is intentional and whether it is important (I don't this on `.md` page), but some lines/words here seem to be italicized.
</blockquote></details>

I think this is a bug in Reviewable rendering. It's not in the source code or in the generated output.

**Review by vy-ton — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/12212)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @stbof and @yuzefovich)

___
*[v21.2/transactions.md, line 270 at r2](https://reviewable.io/reviews/cockroachdb/docs/12212#-MoEw_zh2qmxfhZHne9F:-MoFSzZ2ExJSYDRGj2hT:b-ypleqc) ([raw file](https://github.com/cockroachdb/docs/blob/21aa808d413226b3d23686a9307643de57f5a572/v21.2/transactions.md#L270)):*
<details><summary><i>Previously, stbof (Stephanie Bodoff) wrote…</i></summary><blockquote>

I'm not sure I understand what you are asking for. The setting names are the same for clusters and sessions. Should I drop cluster settings?
</blockquote></details>

Session setting names remove the `sql.defaults.` prefix. Its just `transaction_rows_writter_err`
___
*[v21.2/transactions.md, line 284 at r2](https://reviewable.io/reviews/cockroachdb/docs/12212#-MoEw0TO6_vMqcp9Y1_Z:-MoFTe7Q8Ppu0sdW0Ei4:b-qcwcym) ([raw file](https://github.com/cockroachdb/docs/blob/21aa808d413226b3d23686a9307643de57f5a572/v21.2/transactions.md#L284)):*
> Should I add a new section to the transactions page that describes explicit and implicit transactions and describes the auto commit behavior for the latter?

We have existing docs for explicit/implicit transactions. @stbof Let's file a separate docs issue with the context Yahor shared for future documentation/prioritization.

We can keep this alert with something more general. Thoughts?
```
Enabling `transaction_rows_read_err` disables a performance optimization for mutation statements in implicit transactions where CockroachDB can auto-commit without additional network round trips.
```
@yuzefovich Is the `read_err` the only instance where the auto-commit optimization is disabled?

**Review by yuzefovich — COMMENTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/12212)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @stbof, @vy-ton, and @yuzefovich)

___
*[v21.2/transactions.md, line 284 at r2](https://reviewable.io/reviews/cockroachdb/docs/12212#-MoEw0TO6_vMqcp9Y1_Z:-MoFcba829eTTJT2XAid:b-vuosnp) ([raw file](https://github.com/cockroachdb/docs/blob/21aa808d413226b3d23686a9307643de57f5a572/v21.2/transactions.md#L284)):*
I agree with Vy - her suggestion looks good to me, and we should document separately the auto-commit optimization more.

> Is the `read_err` the only instance where the auto-commit optimization is disabled?

Yep (neither `write_err` nor `_log` guardrails disable the optimization).

**Review by yuzefovich — APPROVED** [review]
> Reviewed 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/12212)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @stbof and @vy-ton)

**Review by vy-ton — APPROVED** [review]

**Review by ianjevans — APPROVED** [review]

**Pull request merged**